core/access/peer.go

```diff
@@ -0,0 +1,181 @@
+type getReceiptsFn func([]common.Hash) error
+type getProofsFn func([]*ProofReq) error
+
+// Peer stores ODR-specific information about LES peers that are able to serve requests
+type Peer struct {
+	id   string      // Unique identifier of the peer
+	head common.Hash // Hash of the peers latest known block
+
+	rep int32 // Simple peer reputation
+
+	GetBlockBodies getBlockBodiesFn
+	GetNodeData    getNodeDataFn
+	GetReceipts    getReceiptsFn
+	GetProofs      getProofsFn
+
+	version int // LES protocol version number
+}
+
+// newPeer creates a new ODR peer
+func newPeer(id string, version int, head common.Hash, getBlockBodies getBlockBodiesFn, getNodeData getNodeDataFn, getReceipts getReceiptsFn, getProofs getProofsFn) *Peer {
+	return &Peer{
+		id:             id,
+		head:           head,
+		GetBlockBodies: getBlockBodies,
+		GetNodeData:    getNodeData,
+		GetReceipts:    getReceipts,
+		GetProofs:      getProofs,
+		version:        version,
+	}
+}
+
+// Id returns a peer's ID string
+func (p *Peer) Id() string {
+	return p.id
+}
+
+
+// Promote increases the peer's reputation
+func (p *Peer) Promote() {
+	atomic.AddInt32(&p.rep, 1)
+}
+
+// Demote decreases the peer's reputation or leaves it at 0.
+func (p *Peer) Demote() {
+	for {
+		// Calculate the new reputation value
+		prev := atomic.LoadInt32(&p.rep)
+		next := prev / 2
+
+		// Try to update the old value
+		if atomic.CompareAndSwapInt32(&p.rep, prev, next) {
+			return
+		}
+	}
+}
+
+// peerSet represents the collection of active peer participating in the block
+// download procedure.
+type peerSet struct {
+	peers map[string]*Peer
+	lock  sync.RWMutex
+}
+
+// newPeerSet creates a new peer set top track the active download sources.
+func newPeerSet() *peerSet {
+	return &peerSet{
+		peers: make(map[string]*Peer),
+	}
+}
+
+// Register injects a new peer into the working set, or returns an error if the
+// peer is already known.
+func (ps *peerSet) Register(p *Peer) error {
+	ps.lock.Lock()
+	defer ps.lock.Unlock()
+
+	if _, ok := ps.peers[p.id]; ok {
+		return errAlreadyRegistered
+	}
+	ps.peers[p.id] = p
+	return nil
+}
+
+// Unregister removes a remote peer from the active set, disabling any further
+// actions to/from that particular entity.
+func (ps *peerSet) Unregister(id string) error {
+	ps.lock.Lock()
+	defer ps.lock.Unlock()
+
+	if _, ok := ps.peers[id]; !ok {
+		return errNotRegistered
+	}
+	delete(ps.peers, id)
+	return nil
+}
+
+// Peer retrieves the registered peer with the given id.
+func (ps *peerSet) Peer(id string) *Peer {
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+
+	return ps.peers[id]
+}
+
+// Len returns if the current number of peers in the set.
+func (ps *peerSet) Len() int {
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+
+	return len(ps.peers)
+}
+
+// AllPeers retrieves a flat list of all the peers within the set.
+func (ps *peerSet) AllPeers() []*Peer {
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+
+	list := make([]*Peer, 0, len(ps.peers))
```

Comment:
This is actually fine since `append` does not grow the array. Just the slice and there's no additional overhead of realloc every time `append` is called.